modules/gdscript/gdscript_analyzer.cpp

```diff
@@ -1234,8 +1234,8 @@ void GDScriptAnalyzer::resolve_assert(GDScriptParser::AssertNode *p_assert) {
 	reduce_expression(p_assert->condition);
 	if (p_assert->message != nullptr) {
 		reduce_expression(p_assert->message);
-		if (!p_assert->message->is_constant || p_assert->message->reduced_value.get_type() != Variant::STRING) {
-			push_error(R"(Expected constant string for assert error message.)", p_assert->message);
+		if (p_assert->message->get_datatype().builtin_type != Variant::STRING) {
```

Comment:
It still needs a proper check for the type and if it has a type at all, like this:

```suggestion
		if (!p_assert->message->get_datatype().has_no_type() && (p_assert->message->get_datatype().kind != GDScriptParser::DataType::BUILTIN || p_assert->message->get_datatype().builtin_type != Variant::STRING)) {
```

Reply:
If the goal is to not raise an error without the certainty that it's not a string, I would like to suggest this instead:
```suggestion
if (p_assert->message->get_datatype().has_is_hard_type() && (p_assert->message->get_datatype().kind != GDScriptParser::DataType::BUILTIN || p_assert->message->get_datatype().builtin_type != Variant::STRING)) {
```
since for the `INFERRED` case it would still be dynamic and might in fact be a string regardless of the inferred type.

Reply:
@cdemirer The PR author didn't update it since May 2021 so it might be worth opening a new one with your suggestion.

Reply:
Actually I'm not sure about the desired behavior here... I put some thoughts under the original issue (#47157).